pkg/controller/podautoscaler/horizontal.go

```diff
@@ -315,7 +327,8 @@ func (a *HorizontalController) computeReplicasForMetrics(ctx context.Context, hp
 		return 0, "", statuses, time.Time{}, fmt.Errorf("invalid metrics (%v invalid out of %v), first error is: %v", invalidMetricsCount, len(metricSpecs), invalidMetricError)
 	}
 	setCondition(hpa, autoscalingv2.ScalingActive, v1.ConditionTrue, "ValidMetricFound", "the HPA was able to successfully calculate a replica count from %s", metric)
-	return replicas, metric, statuses, timestamp, nil
+
+	return replicas, metric, statuses, timestamp, fmt.Errorf("invalid metrics (%v invalid out of %v), first error is: %v", invalidMetricsCount, len(metricSpecs), invalidMetricError)
```

Comment:
The "first error" is a bit of a lie, because there could be spec error first which is later overridden by internal error.

Now that I think of it perhaps overriding spec with internal is bringing more confusion than value. Perhaps it's good enough and simpler to report in metrics always the first error, regardless if it's spec or internal.

Reply:
For the current impl, "unknown metric source type" is only the spec error we can have here. And actually such a invalid metric source will be validated in the validation of api-server, meaning no spec error can reach here.
https://github.com/kubernetes/kubernetes/blob/04d52a56fd40a72ad9a01bc5b692600fb5ad0097/pkg/apis/autoscaling/validation/validation.go#L327
So, we can ignore spec error case here either way.